src/dataflow/src/render/delta_join.rs

```diff
@@ -232,70 +262,84 @@ where
                                             let mut bound_expr = input_mapper
                                                 .find_bound_expr(expr, &bound_inputs, &equivalences)
                                                 .expect("Expression in join plan is not bound at time of use");
-
-                                            bound_expr.visit_mut(&mut |e| if let ScalarExpr::Column(c) = e {
-                                                *c = source_columns.iter().position(|x| x == c).expect("Did not find bound column in source_columns");
-                                            });
+                                            // Rewrite column references to physical locations.
+                                            bound_expr.permute_map(&column_map);
                                             bound_expr
                                         })
                                         .collect::<Vec<_>>();
 
-                                    // We should extract each element of `next_keys` from `equivalences`,
-                                    // as each *should* now be a redundant constraint. We do this so that
-                                    // the demand analysis does not require these columns be produced.
+                                    // Remove each element of `next_keys` from `equivalences`, so that we
+                                    // avoid redundant predicate work. This removal also paves the way for
+                                    // more precise "demand" information going forward.
                                     for equivalence in equivalences.iter_mut() {
                                         equivalence.retain(|expr| !next_key_rebased.contains(expr));
                                     }
                                     equivalences.retain(|e| e.len() > 1);
 
-                                    // TODO: Investigate demanded columns as in DifferentialLinear join.
```

Comment:
not sure if you meant to discard this todo

Reply:
Good point.

Reply:
Wait. I was confused sorry. Yes, I think I meant to delete it (demanded columns are handled in `MyClosure` now, I think).